### PR TITLE
NAS-108842 / 21.02 / Custom Catalogs Support

### PR DESCRIPTION
**Pull request by sonicaj**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/catalogs_linux/update.py [hunk 1]
```diff
@@ -1,47 +1,117 @@
...
+        await self.middleware.call('datastore.insert', self._config.datastore, data)
+
+        asyncio.ensure_future(self.middleware.call('catalog.sync', data['label']))
+
+        return await self.get_instance(data['label'])
+
+    @accepts(
+        Str('id'),
+    )
+    def do_delete(self, id):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/catalogs_linux/update.py:106]: So you're allowing the OFFICIAL catalog to be deleted?

Seems we should have another attribute in the db (`builtin`) and do not allow builtin to be changed?

File: src/middlewared/middlewared/alert/source/catalogs.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from middlewared.alert.base import AlertClass, AlertCategory, AlertLevel, Alert, OneShotAlertClass
+
+
+class CatalogNotHealthyAlertClass(AlertClass, OneShotAlertClass):
+    deleted_automatically = False
+    level = AlertLevel.WARNING
+    category = AlertCategory.APPLICATIONS
+    title = 'Catalog Not Healthy'
+    text = '%(apps)s Applications in %(catalog)s Catalog are not healthy'
```
Comment (william-gr) [inline src/middlewared/middlewared/alert/source/catalogs.py:9]: Think our text usually ends with a dot?

File: src/middlewared/middlewared/plugins/catalogs_linux/items.py [hunk 3]
```diff
@@ -38,9 +38,11 @@ def items(self, label, options):
         elif not os.path.exists(catalog['location']):
             self.middleware.call_sync('catalog.update_git_repository', catalog, True)
 
+        self.middleware.call_sync('alert.oneshot_delete', 'CatalogNotHealthy', f'"{label}"')
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/catalogs_linux/items.py:41]: Why a f-string?

Comment (william-gr) [reply src/middlewared/middlewared/plugins/catalogs_linux/items.py:41]: Ah, the quotes, nevermind, but why quotes?

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/catalogs_linux/items.py]: So basically when we create an alert specifying `key`, it json dumps the value which results in the key having quotes and thus we need to do the same during deletion. But if we use `_key` instead, that does not happen and i changed that. Thank you for pointing it out :)

**Review by william-gr — APPROVED** [review]

**Pull request merged**